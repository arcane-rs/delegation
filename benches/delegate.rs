use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
