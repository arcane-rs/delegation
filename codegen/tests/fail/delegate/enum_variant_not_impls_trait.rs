use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

#[delegate(derive(AsStr))]
enum Name {
    First(String),
}

fn main() {
    let name = Name::First("John".to_string());
    assert_eq!(name.as_str(), "John");
}
