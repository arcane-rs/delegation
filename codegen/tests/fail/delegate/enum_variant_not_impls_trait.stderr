error[E0277]: the trait bound `String: AsStr` is not satisfied
 --> tests/fail/delegate/enum_variant_not_impls_trait.rs:8:1
  |
8 | #[delegate(derive(AsStr))]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `AsStr` is not implemented for `String`
  |
  = help: the following other types implement trait `AsStr`:
            Name
            delegation::private::Wrapper<__Delegate>
note: required for `Either<String, Void>` to implement `__delegate_AsStr__DelegateOwned`
 --> tests/fail/delegate/enum_variant_not_impls_trait.rs:3:1
  |
3 | #[delegate]
  | ^^^^^^^^^^^
4 | trait AsStr {
  |       ^^^^^ unsatisfied trait bound introduced here
note: required for `delegation::private::Wrapper<Name>` to implement `AsStr`
 --> tests/fail/delegate/enum_variant_not_impls_trait.rs:3:1
  |
3 | #[delegate]
  | ^^^^^^^^^^^
4 | trait AsStr {
  |       ^^^^^ unsatisfied trait bound introduced here
  = note: this error originates in the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
